internal/olm/operator/bundle/install.go

```diff
@@ -98,3 +185,137 @@ func (i *Install) setup(ctx context.Context) error {
+	}
+
+	// check if the bundle exists in the index image.
+	if len(bundleDeclConfig.Channels) > 0 && len(bundleDeclConfig.Bundles) > 0 {
+		for _, channel := range imageDeclConfig.Channels {
+			// find the specific channel that the bundle needs to be inserted into.
+			if channel.Name == bundleDeclConfig.Channels[0].Name && channel.Package == bundleDeclConfig.Channels[0].Package {
+				// check if the CSV name is already present in the channel's entries.
+				for _, entry := range channel.Entries {
+					if entry.Name == bundleDeclConfig.Bundles[0].Name {
+						return nil, fmt.Errorf("bundle %q already present in the index image: %s", bundleDeclConfig.Bundles[0].Name, indexImage)
+					}
+				}
+
+				break // we only want to search through the specific channel.
+			}
+		}
+	}
+
+	if len(bundleDeclConfig.Bundles) > 0 && len(bundleDeclConfig.Channels) > 0 && len(bundleDeclConfig.Packages) > 0 {
+		imageDeclConfig.Bundles = append(imageDeclConfig.Bundles, bundleDeclConfig.Bundles[0])
+		imageDeclConfig.Channels = append(imageDeclConfig.Channels, bundleDeclConfig.Channels[0])
+		imageDeclConfig.Packages = append(imageDeclConfig.Packages, bundleDeclConfig.Packages[0])
+
+		if len(bundleDeclConfig.Others) > 0 {
+			imageDeclConfig.Others = append(imageDeclConfig.Others, bundleDeclConfig.Others[0])
+		}
+
+		log.Infof("Inserted the new bundle %q into the index image: %s", bundleDeclConfig.Bundles[0].Name, indexImage)
+	}
+
+	return imageDeclConfig, nil
+}
+
+// createFBC generates an FBC by creating bundle, package and channel blobs.
+func (f *FBCContext) createFBC(ctx context.Context) (*declarativeconfig.DeclarativeConfig, error) {
+	var (
+		declcfg        *declarativeconfig.DeclarativeConfig
+		declcfgpackage *declarativeconfig.Package
+		err            error
+	)
+
+	// Rendering the bundle image into a declarative config format.
+	log.SetOutput(ioutil.Discard)
+	render := action.Render{
+		Refs: f.Refs,
+	}
+
+	// generate bundles by rendering the bundle objects.
+	declcfg, err = render.Run(ctx)
+	if err != nil {
+		log.Errorf("Error in rendering the bundle image: %v", err)
+		return nil, err
+	}
+	log.SetOutput(os.Stdout)
+
+	// Ensuring a valid bundle size.
+	if len(declcfg.Bundles) < 0 {
+		return nil, fmt.Errorf("bundles rendered are less than 0 for the bundle image")
+	}
+
+	// init packages.
+	init := action.Init{
+		Package:        f.Package,
+		DefaultChannel: f.ChannelName,
+	}
+
+	// generate packages.
+	declcfgpackage, err = init.Run()
+	if err != nil {
+		log.Errorf("Error in generating packages for the FBC: %v", err)
+		return nil, err
+	}
+	declcfg.Packages = []declarativeconfig.Package{*declcfgpackage}
+
+	// generate channels.
+	channel := declarativeconfig.Channel{
+		Schema:  schemaChannel,
+		Name:    f.ChannelName,
+		Package: f.Package,
+		Entries: []declarativeconfig.ChannelEntry{f.ChannelEntry},
+	}
+
+	declcfg.Channels = []declarativeconfig.Channel{channel}
+
+	return declcfg, nil
```

Comment:
Going along with my previous question about the length checks in `addBundleToIndexImage`, perhaps we could declare a new unexported struct like the following that has exactly one of the things we know we're generating:

```go
type bundleDeclcfg struct {
	Package declcfg.Package
	Channel declcfg.Channel
	Bundle declcfg.Bundle
}
```
And then use that for the `declcfg` variable in this function in place of the more lenient `declcfg.DeclarativeConfig` type. I think that would remove the need for the length checks in `addBundleToIndexImage`.

The only catch is that there would need to be some (admittedly very simple) code to convert that to a `declcfg.DeclarativeConfig` struct in the case where we don't have an existing index image.